Google.Api.Gax.Grpc/RpcExceptionExtensions.cs

```diff
@@ -0,0 +1,107 @@
+        /// </summary>
+        /// <param name="ex">The RPC exception to retrieve details from. Must not be null.</param>
+        /// <returns>The <see cref="Status"/> message specified in the exception, or null
+        /// if there is no such information.</returns>
+        public static Rpc.Status GetRpcStatus(this RpcException ex) =>
+            DecodeTrailer(GetTrailer(ex, StatusDetailsTrailerName), Status.Parser);
+
+        /// <summary>
+        /// Retrieves the <see cref="BadRequest"/> message containing extended error information
+        /// from the trailers in an <see cref="RpcException"/>, if present.
+        /// </summary>
+        /// <param name="ex">The RPC exception to retrieve details from. Must not be null.</param>
+        /// <returns>The <see cref="BadRequest"/> message specified in the exception, or null
+        /// if there is no such information.</returns>
+        public static BadRequest GetBadRequest(this RpcException ex) => GetStatusDetail<BadRequest>(ex);
+
+        /// <summary>
+        /// Retrieves the <see cref="ErrorInfo"/> message containing extended error information
+        /// from the trailers in an <see cref="RpcException"/>, if present.
+        /// </summary>
+        /// <param name="ex">The RPC exception to retrieve details from. Must not be null.</param>
+        /// <returns>The <see cref="ErrorInfo"/> message specified in the exception, or null
+        /// if there is no such information.</returns>
+        public static ErrorInfo GetErrorInfo(this RpcException ex) => GetStatusDetail<ErrorInfo>(ex);
+
+        /// <summary>
+        /// Retrieves the error details of type <typeparamref name="T"/> from the <see cref="Status"/>
+        /// message associated with an <see cref="RpcException"/>, if any.
+        /// </summary>
+        /// <typeparam name="T">The message type to decode from within the error details.</typeparam>
+        /// <param name="ex">The RPC exception to retrieve details from. Must not be null.</param>
+        /// <returns></returns>
+        public static T GetStatusDetail<T>(this RpcException ex) where T : class, IMessage<T>, new()
+        {
+            var status = GetRpcStatus(ex);
+            if (status is null)
+            {
+                return null;
+            }
+            var expectedName = new T().Descriptor.FullName;
+            var any = status.Details.FirstOrDefault(a => Any.GetTypeName(a.TypeUrl) == expectedName);
+            if (any is null)
+            {
+                return null;
+            }
+            try
+            {
+                return any.Unpack<T>();
+            }
+            catch
+            {
+                // If the message is malformed, just report there's no information.
+                return null;
+            }
+        }
+
+        private static Metadata.Entry GetTrailer(RpcException ex, string key)
+        {
+            GaxPreconditions.CheckNotNull(ex, nameof(ex));
```

Comment:
Try to understand the behavior of `CheckNotNull`. It is possible gRPC metadata do not have key `grpc-status-details-bin`, then we should return the original error instead throw new error.

Reply:
`CheckNotNull` is checking that you haven't passed a null value for `ex` - that's all. Basically, this should fail:

```csharp
RpcException ex = null;
ex.GetErrorInfo();
```

But (as per the tests) it's absolutely fine to call it on an exception that doesn't have a trailer. There's no notion of "return the original error" though - the caller already has the `RpcException`; that's never returned. The only thing that's returned is extra information from within that exception.

Reply:
Perfect. Thanks for clarification. 👍